elfpy/utils/apeworx_integrations.py

```diff
@@ -397,50 +407,38 @@ def get_wallet_from_trade_history(
-                        value * trade_history.loc[specific_trade, "share_price"]
-                    )
-                open_share_price = int(sum_product_of_open_share_price_and_value / sum_value)
-                assert abs(balance - sum_value) <= tolerance, "weighted average open share price calculation is wrong"
-                logging.debug("calculated weighted average open share price of %s", open_share_price)
                 previous_balance = wallet.shorts[mint_time].balance if mint_time in wallet.shorts else 0
-                new_balance = previous_balance + FixedPoint(scaled_value=balance)
+                delta_balance = FixedPoint(scaled_value=balance)
+                new_balance = previous_balance + delta_balance
                 if new_balance == 0:
-                    # remove key of "mint_time" from the "wallet.shorts" dict
                     wallet.shorts.pop(mint_time, None)
-                else:  # position balance is not zero
-                    wallet.shorts.update(
-                        {
-                            mint_time: Short(
-                                balance=new_balance,
-                                open_share_price=FixedPoint(scaled_value=open_share_price),
-                            )
-                        }
-                    )
-                    logging.debug(
-                        "storing in wallet as %s",
-                        {
-                            mint_time: Short(
-                                balance=new_balance, open_share_price=FixedPoint(scaled_value=open_share_price)
-                            )
-                        },
-                    )
+                else:  # we either do a marginal update or full weighted average for open share price
+                    if add_to_existing_wallet:  # marginal update, assuming trade_history contains exactly ONE trade
+                        previous_share_price = (
+                            wallet.shorts[mint_time].open_share_price if mint_time in wallet.shorts else 0
+                        )
+                        delta_open_share_price = FixedPoint(scaled_value=int(trade_history.share_price.iloc[0]))
+                        # weighted average update: new_y = ( old_x * old_y + new_x * new_y ) / (old_x + new_x)
+                        updated_open_share_price = (
+                            previous_balance * previous_share_price + delta_balance * delta_open_share_price
+                        ) / new_balance
+                        updated_open_share_price = FixedPoint(scaled_value=int(updated_open_share_price))
+                    else:  # weighted average across a bunch of trades, assuming trade_history contains EVERY trade
+                        value = trades_df["value"] * np.where(trades_df["from"] == address, -1, 1)
+                        updated_open_share_price = FixedPoint(int(np.average(trades_df["share_price"], weights=value)))
```

Comment:
are you sure you want to cast this to `int`? The weighted average is going to be a float, so casting to `int` will perform a truncation of the decimal portion. The better way to convert to int is to [round half to even](https://en.wikipedia.org/wiki/Rounding#:~:text=Rounding%20half%20to%20even%5Bedit%5D). Or, assuming you want to keep the decimal information, why not cast it to float and pass it directly to FixedPoint? e.g. `FixedPoint(np.average(...).item())` should work.

Reply:
changed to do so in a helper function with error-checking